File: docs/sdd/unified-qtquick-controls/audit/palette-transition.cpp
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: 2026 Andrii L <lebeden@gmail.com>

// Standalone diagnostic: collection errors fail here; P03 assertions are separate.
#include <QFile>
#include <QGuiApplication>
#include <QHoverEvent>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QMetaEnum>
#include <QQmlComponent>
#include <QQmlContext>
#include <QQmlEngine>
#include <QQuickItem>
#include <QQuickWindow>
#include <QSaveFile>
#include <QSet>
#include <QtQuick/private/qquickitem_p.h>
#include <QtQuick/private/qquickpalette_p.h>
#include <QtQuick/private/qquickwindow_p.h>
#include <QtTest/QTest>
#include <qpa/qwindowsysteminterface.h>

#include <memory>

namespace {
QJsonObject paletteState(const QPalette& palette) {
  QJsonObject result{{"resolveMask", QString::number(palette.resolveMask(), 16)}};
  const auto roles = QMetaEnum::fromType<QPalette::ColorRole>();
  for (auto group : {QPalette::Active, QPalette::Inactive, QPalette::Disabled}) {
    QJsonObject colors;
    for (int index = 0; index < QPalette::NColorRoles; ++index)
      colors[QString::fromLatin1(roles.valueToKey(index))] =
          palette.color(group, static_cast<QPalette::ColorRole>(index)).name(QColor::HexArgb);
    result[QString::number(group)] = colors;
  }
  return result;
}
void require(bool condition, const char* message) {
  if (!condition) qFatal("%s", message);
}
void writeAppearance(bool light) {
  QFile source(qEnvironmentVariable("HOLONIGHT_APPEARANCE_FILE"));
  require(source.open(QIODevice::ReadOnly), "Cannot read isolated appearance file");
  auto contents = source.readAll();
  contents.replace(light ? "holonight-dark" : "holonight-light", light ? "holonight-light" : "holonight-dark");
  source.close();
  QSaveFile file(source.fileName());
  require(file.open(QIODevice::WriteOnly), "Cannot open isolated appearance file");
  file.write(contents);
  require(file.commit(), "Cannot commit appearance file");
}
QPalette explicitPalette(bool light) {
  QPalette palette;
  for (auto group : {QPalette::Active, QPalette::Inactive, QPalette::Disabled}) {
    for (auto role : {QPalette::Window, QPalette::Base, QPalette::Button, QPalette::AlternateBase})
      palette.setColor(group, role, light ? QColor("#eeeeee") : QColor("#222222"));
    for (auto role : {QPalette::Text, QPalette::WindowText, QPalette::ButtonText})
      palette.setColor(group, role, light ? QColor("#222222") : QColor("#eeeeee"));
  }
  return palette;
}
QJsonObject objectState(QObject* object) {
  QJsonObject state{{"class", object->metaObject()->className()},
                    {"name", object->objectName()},
                    {"identity", QString::number(reinterpret_cast<quintptr>(object), 16)}};
  if (auto* context = qmlContext(object)) state["origin"] = context->baseUrl().toString();
  // Reading Item.palette can allocate it. Stability observation must not create
  // an inheritance boundary; allocation is a separate explicit experiment.
  bool allocated = true;
  if (!qEnvironmentVariableIsEmpty("UQC_STABILITY_TRIGGER")) {
    if (auto* item = qobject_cast<QQuickItem*>(object)) allocated = QQuickItemPrivate::get(item)->providesPalette();
    if (auto* window = qobject_cast<QQuickWindow*>(object))
      allocated = QQuickWindowPrivate::get(window)->providesPalette();
  }
  state["paletteAllocated"] = allocated;
  if (allocated) {
    if (auto* palette = object->property("palette").value<QQuickPalette*>()) {
      state["palette"] = paletteState(palette->toQPalette());
      state["currentColorGroup"] = int(palette->currentColorGroup());
    }
  }
  for (const auto* name :
       {"enabled", "visible", "color", "active", "hovered", "activeFocus", "appearanceRevision", "appearanceBase"}) {
    const auto value = object->property(name);
    if (value.isValid())
      state[name] = value.metaType() == QMetaType::fromType<QColor>()
                        ? QJsonValue(value.value<QColor>().name(QColor::HexArgb))
                        : QJsonValue::fromVariant(value);
  }
  return state;
}
void walk(QQuickItem* item, QJsonArray& objects, QSet<QQuickItem*>& visited, const QImage& image, qreal dpr) {
  if (visited.contains(item)) return;
  visited.insert(item);
  auto state = objectState(item);
  if (item->parentItem()) state["parentClass"] = item->parentItem()->metaObject()->className();
  // These fixtures deliberately instantiate backgrounds; observing them is safe here.
  if (!item->objectName().isEmpty()) {
    if (auto* background = item->property("background").value<QQuickItem*>())
      state["background"] = objectState(background);
    const auto point = item->mapToScene(QPointF(item->width() - 25, item->height() / 2)) * dpr;
    if (image.rect().contains(point.toPoint()))
      state["pixel"] = image.pixelColor(point.toPoint()).name(QColor::HexArgb);
  }
  objects.append(state);
  for (auto* child : item->childItems()) walk(child, objects, visited, image, dpr);
}
QJsonObject snapshot(QQuickWindow* window, const QString& output, const QString& phase, const QString& name) {
  int frames = 0;
  const auto connection = QObject::connect(window, &QQuickWindow::frameSwapped, window, [&] { ++frames; });
  window->setProperty("captureRevision", window->property("captureRevision").toInt() + 1);
  window->update();
  require(QTest::qWaitFor([&] { return frames > 0; }, 5000), "Frame transition timeout");
  QObject::disconnect(connection);
  const auto image = window->grabWindow();
  require(!image.isNull(), "Rendered capture unavailable");
  require(image.save(output + "/" + phase + "-" + name + ".png"), "Cannot save capture");
  QJsonArray objects;
  QSet<QQuickItem*> visited;
  walk(window->contentItem(), objects, visited, image, window->devicePixelRatio());
  auto result = objectState(window);
  result["contentItem"] = objectState(window->contentItem());
  result["dpr"] = window->devicePixelRatio();
  result["objects"] = objects;
  return result;
}
// UQC-223 is opt-in and may dispatch only inside a disposable offscreen process.
class Observer : public QObject {
 public:
  QJsonArray events;
  QString operation;
  bool eventFilter(QObject* object, QEvent* event) override {
    switch (event->type()) {
      case QEvent::WindowActivate:
      case QEvent::WindowDeactivate:
      case QEvent::FocusIn:
      case QEvent::FocusOut:
      case QEvent::HoverEnter:
      case QEvent::HoverLeave:
      case QEvent::PaletteChange:
      case QEvent::ApplicationPaletteChange:
        events.append(QJsonObject{
            {"operation", operation}, {"type", int(event->type())}, {"beforeDelivery", objectState(object)}});
        break;
      default:
        break;
    }
    return false;
  }
  void watch(QObject* object) {
    if (auto* item = qobject_cast<QQuickItem*>(object); item && !QQuickItemPrivate::get(item)->providesPalette())
      return;
    if (auto* palette = object->property("palette").value<QQuickPalette*>())
      connect(palette, &QQuickPalette::changed, this, [this, object] {
        events.append(
            QJsonObject{{"operation", operation}, {"signal", "palette.changed"}, {"state", objectState(object)}});
      });
    if (auto* window = qobject_cast<QQuickWindow*>(object))
      connect(window, &QWindow::activeChanged, this, [this, window] {
        events.append(
            QJsonObject{{"operation", operation}, {"signal", "activeChanged"}, {"state", objectState(window)}});
      });
  }
};
QJsonObject immediate(QQuickWindow* window) {
  auto state = objectState(window);
  QJsonArray objects;
  QSet<QQuickItem*> visited;
  walk(window->contentItem(), objects, visited, QImage{}, window->devicePixelRatio());
  state["objects"] = objects;
  return state;
}
QJsonArray stability(QQmlComponent& component, const QString& output) {
  require(QGuiApplication::platformName() == "offscreen", "Synthetic diagnostics require offscreen");
  Observer observer;
  qGuiApp->installEventFilter(&observer);
  const auto trigger = qEnvironmentVariable("UQC_STABILITY_TRIGGER");
  const bool lightHistory = qEnvironmentVariable("UQC_SECONDARY_HISTORY") == "light";
  const auto overrideScope = qEnvironmentVariable("UQC_OVERRIDE_SCOPE");
  auto create = [&](const QString& name) {
    auto window = std::unique_ptr<QQuickWindow>(qobject_cast<QQuickWindow*>(component.create()));
    if (!window) qFatal("Fixture failed: %s", qPrintable(component.errorString()));
    window->setObjectName(name);
    require(QTest::qWaitFor([&] { return window->isExposed(); }, 5000), "Exposure timeout");
    if (qEnvironmentVariable("UQC_CONTENT_PALETTE") == "read") {
      // Reproduce exactly the retained observer's palette getter, independently
      // of the passive recorder. This intentionally allocates a root palette.
      window->contentItem()->property("palette");
    }
    if (overrideScope == "window" || overrideScope == "control") {
      QObject* target =
          overrideScope == "window" ? static_cast<QObject*>(window.get()) : window->findChild<QObject*>("nativeText");
      auto* palette = target->property("palette").value<QQuickPalette*>();
      palette->setBase(QColor("#bada55"));
      palette->disabled()->setText(QColor("#112233"));
    }
    observer.watch(window.get());
    for (auto* child : window->findChildren<QObject*>())
      if (!child->objectName().isEmpty()) observer.watch(child);
    return window;
  };
  const bool native = qEnvironmentVariable("UQC_BOUNDARY") == "native";
  if (native) QGuiApplication::setPalette(explicitPalette(false));
  if (overrideScope == "application") {
    QPalette palette;
    palette.setColor(QPalette::Base, QColor("#bada55"));
    palette.setColor(QPalette::Disabled, QPalette::Text, QColor("#112233"));
    QGuiApplication::setPalette(palette);
  }
  auto primary = create("main-created-dark");
  const auto initialPrimary = snapshot(primary.get(), output, "construction", "primary");
  std::unique_ptr<QQuickWindow> secondary;
  if (!lightHistory) secondary = create("secondary-created-dark");
  auto transition = [&](bool light) {
    const auto revision = primary->property("appearanceRevision").toInt();
    writeAppearance(light);
    if (native) {
      QGuiApplication::setPalette(explicitPalette(light));
      primary->setProperty("appearanceRevision", revision + 1);
      if (secondary) secondary->setProperty("appearanceRevision", revision + 1);
    }
    require(QTest::qWaitFor([&] { return primary->property("appearanceRevision").toInt() > revision; }, 5000),
            "Appearance revision timeout");
  };
  QJsonArray records;
  auto collect = [&](const QString& phase) {
    QFile appearance(qEnvironmentVariable("HOLONIGHT_APPEARANCE_FILE"));
    require(appearance.open(QIODevice::ReadOnly), "Appearance read failed");
    records.append(QJsonObject{{"phase", phase},
                               {"requestedAppearance", QString::fromUtf8(appearance.readAll())},
                               {"application", paletteState(QGuiApplication::palette())},
                               {"immediatePrimary", immediate(primary.get())},
                               {"immediateSecondary", immediate(secondary.get())},
                               {"primary", snapshot(primary.get(), output, phase, "primary")},
                               {"secondary", snapshot(secondary.get(), output, phase, "secondary")}});
  };
  transition(true);
  const auto beforeSecondary = snapshot(primary.get(), output, "transition", "primary");
  if (lightHistory) secondary = create("secondary-created-light");
  for (bool light : {true, false}) {
    if (!light) transition(false);
    const QString scheme = light ? "light" : "dark-return";
    observer.operation = scheme + "-before";
    collect(observer.operation);
    for (bool enter : {true, false}) {
      observer.operation = scheme + (enter ? "-enter" : "-leave");
      if (trigger == "activation") {
        QWindowSystemInterface::handleFocusWindowChanged(enter ? primary.get() : secondary.get(),
                                                         Qt::ActiveWindowFocusReason);
        QWindowSystemInterface::sendWindowSystemEvents(QEventLoop::AllEvents);
      } else if (trigger == "application-event") {
        QEvent event(QEvent::ApplicationPaletteChange);
        QCoreApplication::sendEvent(qGuiApp, &event);
      } else {
        for (auto* window : {primary.get(), secondary.get()}) {
          for (const auto* name : {"nativeButton", "nativeText", "nativeSpin", "nativeCombo", "disabledText",
                                   "localOverride", "sharedSearch", "sharedCombo"}) {
            auto* item = window->findChild<QQuickItem*>(name);
            if (native && QString::fromLatin1(name).startsWith("shared")) continue;
            require(item, "Missing named control");
            if (trigger == "hover") {
              QHoverEvent event(enter ? QEvent::HoverEnter : QEvent::HoverLeave, QPointF(10, 10), QPointF(10, 10),
                                QPointF(-1, -1));
              QCoreApplication::sendEvent(item, &event);
            } else if (trigger == "resolve") {
              // Mutating an unasserted Link role forces normal palette re-resolution.
              auto* palette = item->property("palette").value<QQuickPalette*>();
              if (enter)
                palette->setLink(QColor("#123456"));
              else
                palette->resetLink();
            } else if (trigger == "enabled" && QString::fromLatin1(name) != "disabledText") {
              item->setEnabled(!enter);
            }
          }
        }
      }
      collect(observer.operation);
    }
  }
  if (overrideScope != "none" && !overrideScope.isEmpty()) {
    observer.operation = "override-reset";
    if (overrideScope == "application")
      QGuiApplication::setPalette(QPalette{});
    else
      for (auto* window : {primary.get(), secondary.get()}) {
        QObject* target =
            overrideScope == "window" ? static_cast<QObject*>(window) : window->findChild<QObject*>("nativeText");
        target->property("palette").value<QQuickPalette*>()->reset();
      }
    QFile reset(output + "/override-reset.json");
    require(reset.open(QIODevice::WriteOnly), "Cannot save override reset");
    reset.write(QJsonDocument(QJsonObject{{"primary", snapshot(primary.get(), output, "reset", "primary")},
                                          {"secondary", snapshot(secondary.get(), output, "reset", "secondary")}})
                    .toJson());
  }
  QFile events(output + "/events.json");
  require(events.open(QIODevice::WriteOnly), "Cannot save events");
  events.write(QJsonDocument(observer.events).toJson());
  auto first = records.first().toObject();
  first["initialPrimary"] = initialPrimary;
  first["beforeSecondaryPrimary"] = beforeSecondary;
  records[0] = first;
  return records;
}
}  // namespace
int main(int argc, char** argv) {
  QGuiApplication app(argc, argv);
  app.setQuitOnLastWindowClosed(false);
  require(argc == 4, "Usage: palette-transition fixture.qml file|application|stability output-directory");
  const bool application = QString::fromLocal8Bit(argv[2]) == "application";
  const auto output = QString::fromLocal8Bit(argv[3]);
  QQmlEngine engine;
  QQmlComponent component(&engine, QUrl::fromLocalFile(QString::fromLocal8Bit(argv[1])));
  auto create = [&]() {
    auto window = std::unique_ptr<QQuickWindow>(qobject_cast<QQuickWindow*>(component.create()));
    if (!window) qFatal("Fixture failed: %s", qPrintable(component.errorString()));
    require(QTest::qWaitFor([&] { return window->isExposed(); }, 5000), "Window exposure timeout");
    return window;
  };
  QJsonArray records;
  if (QString::fromLocal8Bit(argv[2]) == "stability") {
    records = stability(component, output);
  } else {
    auto primary = create();
    auto secondary = create();
    if (application) QGuiApplication::setPalette(explicitPalette(false));
    auto collect = [&](const QString& phase) {
      records.append(QJsonObject{{"phase", phase},
                                 {"application", paletteState(QGuiApplication::palette())},
                                 {"primary", snapshot(primary.get(), output, phase, "primary")},
                                 {"secondary", snapshot(secondary.get(), output, phase, "secondary")}});
    };
    collect("dark");
    for (bool light : {true, false}) {
      const auto previous = primary->property("appearanceBase");
      if (application) {
        QGuiApplication::setPalette(explicitPalette(light));
        require(QTest::qWaitFor([&] { return QGuiApplication::palette() == explicitPalette(light); }, 5000),
                "Application palette transition timeout");
      } else {
        writeAppearance(light);
        require(QTest::qWaitFor([&] { return primary->property("appearanceBase") != previous; }, 5000),
                "Appearance transition timeout");
      }
      collect(light ? "light" : "dark-return");
      if (light) {
        secondary.reset();
        secondary = create();
        collect("light-recreated");
      }
    }
  }
  QFile maps("/proc/self/maps");
  require(maps.open(QIODevice::ReadOnly), "Cannot read mappings");
  QFile mapped(output + "/maps.txt");
  require(mapped.open(QIODevice::WriteOnly), "Cannot save mappings");
  mapped.write(maps.readAll());
  QFile results(output + "/measurements.json");
  require(results.open(QIODevice::WriteOnly), "Cannot save measurements");
  results.write(QJsonDocument(QJsonObject{{"qt", qVersion()}, {"records", records}}).toJson());
}

File: docs/sdd/unified-qtquick-controls/audit/palette-transition.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-3.0-or-later
# SPDX-FileCopyrightText: 2026 Andrii L <lebeden@gmail.com>
"""Collect UQC-221 fixtures, or separately assert native appearance transitions."""

import argparse
import json
import os
from pathlib import Path
import subprocess


APPEARANCE = """
version = 1

[theme]
scheme = "holonight-dark"
accent = "blue"

[typography]
ui_family = "Inter"
ui_size = 12
monospace_family = "JetBrains Mono"
monospace_size = 12
title_family = "Audiowide"
title_size = 10
display_family = "Rajdhani"
display_size = 24

[icons]
theme = "HoloNight"
fallback = "Papirus"
cursor = "default"

[layout]
scale = 1.0

[shape]
style = "inherit"
scale = 1.0
"""


def fixture(window, native, stability=False):
    imports = "import QtQuick\nimport QtQuick.Controls as C\n"
    if not native:
        imports += "import Holonight.Core\nimport Holonight.Controls\n"
    body = """
    id: fixtureRoot
    width: 360; height: 390; visible: true
    objectName: "fixtureWindow"
    property int captureRevision: 0
    Rectangle { width: 2; height: 2; color: fixtureRoot.captureRevision % 2 ? "white" : "black" }
    Column {
        x: 20; y: 20; spacing: 12
        C.TextField { objectName: "nativeText"; width: 300; text: "Native text" }
        C.SpinBox { objectName: "nativeSpin"; width: 300; value: 42 }
        C.ComboBox { objectName: "nativeCombo"; width: 300; model: ["First", "Second"] }
        C.TextField { objectName: "localOverride"; width: 300; text: "Local override"; palette.base: "#bada55" }
        C.TextField { objectName: "disabledText"; width: 300; text: "Disabled"; enabled: false }
"""
    if not native:
        body += """
        HnSearchField { objectName: "sharedSearch"; width: 300; text: "Search" }
        HnIconComboBox { objectName: "sharedCombo"; width: 300; model: ["First", "Second"] }
"""
    body += "    }\n"
    if not native:
        body += "    property color appearanceBase: HoloniightPalette.background\n"
    if stability:
        body = body.replace("height: 390", "height: 550")
        body = body.replace(
            "x: 20; y: 20; spacing: 12",
            'x: 20; y: 20; spacing: 12\n        C.Button { objectName: "nativeButton"; width: 300; text: "Button"; hoverEnabled: true }',
        )
        body += (
            "    property int appearanceRevision: "
            + ("0" if native else "HoloniightPalette.revision")
            + "\n"
        )
    return imports + window + " {\n" + body + "}\n"


def collect(args):
    args.output.mkdir(parents=True, exist_ok=False)
    prefix = args.prefix.resolve()
    cases = (
        [
            (
                args.boundary == "native",
                "Window" if args.boundary != "shared" else "HnApplicationWindow",
            )
        ]
        if args.stability
        else [
            (False, "Window"),
            (False, "HnApplicationWindow"),
            (True, "Window"),
        ]
    )
    for native, window in cases:
        for style in ["Fusion"] if native else ["Fusion", "Holonight"]:
            for mechanism in (
                ["stability"]
                if args.stability
                else (["application"] if native else ["file", "application"])
            ):
                for scale in [args.scale] if args.stability else ["1", "1.25"]:
                    name = (
                        f"{'native' if native else window}-{style}-{mechanism}-{scale}"
                    )
                    case = args.output / name
                    case.mkdir()
                    qml = case / "fixture.qml"
                    qml.write_text(fixture(window, native, bool(args.stability)))
                    appearance = case / "appearance.toml"
                    appearance.write_text(APPEARANCE)
                    env = {
                        k: v
                        for k, v in os.environ.items()
                        if not k.startswith(("QT_", "QML", "HOLONIGHT_", "LD_"))
                    }
                    for key in [
                        "XDG_CONFIG_HOME",
                        "XDG_DATA_HOME",
                        "XDG_CACHE_HOME",
                        "XDG_STATE_HOME",
                        "XDG_RUNTIME_DIR",
                    ]:
                        path = case / key.lower()
                        path.mkdir(mode=0o700)
                        env[key] = str(path)
                    env.update(
                        QT_QPA_PLATFORM="offscreen",
                        QT_QUICK_BACKEND="software",
                        QT_QUICK_CONTROLS_STYLE=style,
                        QT_SCALE_FACTOR=scale,
                        QT_QPA_PLATFORMTHEME=""
                        if native or args.no_platform_theme
                        else "holonight",
                        UQC_STABILITY_TRIGGER=args.stability or "",
                        UQC_SECONDARY_HISTORY=args.history,
                        UQC_CONTENT_PALETTE=args.content_palette,
                        UQC_BOUNDARY=args.boundary,
                        UQC_OVERRIDE_SCOPE=args.override_scope,
                        QT_PLUGIN_PATH=""
                        if native
                        else str(prefix / "lib/qt6/plugins"),
                        QML_IMPORT_PATH="" if native else str(prefix / "lib/qt6/qml"),
                        LD_LIBRARY_PATH="" if native else str(prefix / "lib"),
                        HOLONIGHT_APPEARANCE_FILE=str(appearance),
                        QT_FORCE_STDERR_LOGGING="1",
                        QML_IMPORT_TRACE="1",
                        QT_DEBUG_PLUGINS="1",
                    )
                    if args.stability and style == "Holonight":
                        env.pop("QT_QUICK_CONTROLS_STYLE")
                    if args.observer:
                        env["LD_PRELOAD"] = str(args.observer.resolve())
                        env["HOLONIGHT_PALETTE_DIAGNOSTICS"] = "1"
                    (case / "experiment.json").write_text(
                        json.dumps(
                            {
                                "trigger": args.stability,
                                "observer": str(args.observer.resolve())
                                if args.observer
                                else None,
                                "history": args.history,
                                "content_palette": args.content_palette,
                                "boundary": args.boundary,
                                "override_scope": args.override_scope,
                                "platform_theme": env["QT_QPA_PLATFORMTHEME"],
                                "style_override": env.get("QT_QUICK_CONTROLS_STYLE"),
                                "scale": scale,
                            },
                            indent=2,
                        )
                        + "\n"
                    )
                    bus = case / "bus.conf"
                    bus.write_text(
                        "<busconfig><type>session</type><listen>unix:tmpdir=/tmp</listen>"
                        '<policy context="default"><allow user="*"/><allow send_destination="*"/>'
                        '<allow receive_sender="*"/><allow own="*"/></policy></busconfig>'
                    )
                    command = [
                        "bwrap",
                        "--bind",
                        "/",
                        "/",
                        "--dev",
                        "/dev",
                        "--die-with-parent",
                    ]
                    for host in [
                        "/usr/lib/qt6/qml/Holonight",
                        "/usr/local/lib/qt6/qml/Holonight",
                    ]:
                        if Path(host).exists():
                            command += ["--tmpfs", host]
                    host_theme = "/usr/lib/qt6/plugins/platformthemes/libqholonight.so"
                    if Path(host_theme).exists():
                        empty = case / "hidden-plugin"
                        empty.touch()
                        command += ["--ro-bind", str(empty), host_theme]
                    command += [
                        "dbus-run-session",
                        "--config-file=" + str(bus),
                        "--",
                        str(args.executable.resolve()),
                        str(qml),
                        mechanism,
                        str(case),
                    ]
                    (case / "command.json").write_text(
                        json.dumps(command, indent=2) + "\n"
                    )
                    with (case / "launch.log").open("w") as log:
                        result = subprocess.run(
                            command,
                            env=env,
                            stdout=log,
                            stderr=subprocess.STDOUT,
                            timeout=40,
                        )
                    if result.returncode:
                        raise RuntimeError(
                            f"{name}: collection failed ({result.returncode}); see launch.log"
                        )
                    maps = (case / "maps.txt").read_text()
                    if args.observer:
                        assert str(args.observer.resolve()) in maps
                        assert "HN_PALETTE " in (case / "launch.log").read_text()
                    libraries = sorted(
                        {
                            line.split()[-1]
                            for line in maps.splitlines()
                            if "holonight" in line.lower()
                            and ".so" in line
                            and (
                                not args.observer
                                or line.split()[-1] != str(args.observer.resolve())
                            )
                        }
                    )
                    if style == "Fusion":
                        assert "libQt6QuickControls2Fusion.so" in maps
                    if native:
                        assert not libraries, libraries
                        assert "/platformthemes/" not in maps
                    else:
                        assert any("libqholonight.so" in lib for lib in libraries) == (
                            not args.no_platform_theme
                        )
                        assert libraries and all(
                            str(prefix) in lib for lib in libraries
                        ), libraries
                    data = json.loads((case / "measurements.json").read_text())
                    assert all(
                        r[w]["dpr"] == float(scale)
                        for r in data["records"]
                        for w in ["primary", "secondary"]
                    )
                    (case / "isolation.json").write_text(
                        json.dumps({"libraries": libraries, "verified": True}, indent=2)
                    )
                    print(name, "collected; mappings and DPR verified", flush=True)


def check(args):
    failures = []
    selected = 0
    for path in sorted(args.output.glob("*/measurements.json")):
        case_window = path.parent.name.split("-", 1)[0]
        if args.window != "all" and case_window != args.window:
            continue
        is_file = "-file-" in path.parent.name
        if (args.scope == "p03") != is_file:
            continue
        if args.scope == "external" and not path.parent.name.startswith("native-"):
            continue
        selected += 1
        records = json.loads(path.read_text())["records"]
        assert [r["phase"] for r in records] == [
            "dark",
            "light",
            "light-recreated",
            "dark-return",
        ]
        for window in ["primary", "secondary"]:
            states = [
                {o["name"]: o for o in r[window]["objects"] if o["name"]}
                for r in records
            ]
            for name in ["nativeText", "nativeSpin", "nativeCombo", "disabledText"]:
                role = "Button" if name == "nativeCombo" else "Base"
                group = "1" if name == "disabledText" else "0"
                colors = [s[name]["palette"][group][role] for s in states]
                # Recreated and existing controls must agree; dark must round-trip.
                passed = colors[0] != colors[1] and colors[0] == colors[3]
                if args.scope != "positive" or case_window == "HnApplicationWindow":
                    passed = passed and colors[1] == colors[2]
                if not passed:
                    failures.append(f"{path.parent.name}/{window}/{name}: {colors}")
            if args.scope == "p03" and case_window == "HnApplicationWindow":
                for name in [
                    "nativeText",
                    "nativeSpin",
                    "nativeCombo",
                    "disabledText",
                    "sharedSearch",
                    "sharedCombo",
                ]:
                    for label, values in [
                        (
                            "background",
                            [s[name]["background"]["color"] for s in states],
                        ),
                        ("pixel", [s[name]["pixel"] for s in states]),
                    ]:
                        if not (
                            values[0] != values[1]
                            and values[1] == values[2]
                            and values[0] == values[3]
                        ):
                            failures.append(
                                f"{path.parent.name}/{window}/{name}/{label}: {values}"
                            )
                for name in ["sharedSearch", "sharedCombo"]:
                    colors = [s[name]["palette"]["0"]["Base"] for s in states]
                    assert (
                        colors[0] != colors[1]
                        and colors[1] == colors[2]
                        and colors[0] == colors[3]
                    )
            assert all(
                s["localOverride"]["palette"]["0"]["Base"] == "#ffbada55"
                for s in states
            )
    expected = {
        "all": {"p03": 8, "positive": 10, "external": 2},
        "HnApplicationWindow": {"p03": 4, "positive": 4, "external": 0},
        "Window": {"p03": 4, "positive": 4, "external": 0},
        "native": {"p03": 0, "positive": 2, "external": 2},
    }[args.window][args.scope]
    assert expected, "The selected scope and window have no cases"
    assert selected == expected, f"Expected {expected} cases, found {selected}"
    for failure in failures:
        print(args.scope.upper(), "FAIL", failure)
    print(
        f"{len(failures)} palette/render transition failures; collection is independent"
    )
    return bool(failures)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("action", choices=["collect", "assert"])
    parser.add_argument("--output", required=True, type=lambda p: Path(p).resolve())
    parser.add_argument(
        "--scope", choices=["p03", "positive", "external"], default="p03"
    )
    parser.add_argument(
        "--window",
        choices=["all", "HnApplicationWindow", "Window", "native"],
        default="all",
        help="Filter assertions without changing collection or external boundary evidence",
    )
    parser.add_argument(
        "--stability",
        choices=["activation", "hover", "resolve", "application-event", "enabled"],
    )
    parser.add_argument("--history", choices=["light", "dark"], default="light")
    parser.add_argument("--scale", choices=["1", "1.25"], default="1.25")
    parser.add_argument("--no-platform-theme", action="store_true")
    parser.add_argument(
        "--content-palette", choices=["passive", "read"], default="passive"
    )
    parser.add_argument(
        "--boundary", choices=["shared", "plain", "native"], default="shared"
    )
    parser.add_argument(
        "--override-scope",
        choices=["none", "application", "window", "control"],
        default="none",
    )
    parser.add_argument(
        "--observer", type=Path, help="Load the actual observer shared library"
    )
    parser.add_argument("--prefix", type=Path)
    parser.add_argument("--executable", type=Path)
    args = parser.parse_args()
    if args.action == "collect":
        if not args.prefix or not args.executable:
            parser.error("collect requires --prefix and --executable")
        collect(args)
        return 0
    return check(args)


if __name__ == "__main__":
    raise SystemExit(main())
